### PR TITLE
First implementation of best location for legends

### PR DESCRIPTION
**Pull request by benoit9126**
Hi,
Here is a first trial to implement the "best" legend location. If you are interested... Any improvements are most welcome.

### CODE REVIEW

**Comment by benoit9126** [comment]
Two tests are failing:
- the one I have created to test different legend locations [png](http://chunk.io/f/cc42b2e941b9492ab01b72d0ea93aecb) [pdf](http://chunk.io/f/62cb2abaa2b44730869d775a41de1411)
- text_overlay [png](http://chunk.io/f/76432e2e53c64fcc8524203c91a5f727) [pdf](http://chunk.io/f/98af804a604f47999849fcec1859a08d)

**Comment by benoit9126** [comment]
@nschloe What do you think of this implementation?

**Comment by nschloe** [comment]
Don't worry about linthub too much, it's not a required check anyways.

File: matplotlib2tikz/legend.py [hunk 1]
```diff
@@ -1,6 +1,9 @@
 # -*- coding: utf-8 -*-
 #
 import warnings
+
+import numpy as np
```
Comment (nschloe) [inline matplotlib2tikz/legend.py]: No `as` imports please.

**Comment by nschloe** [comment]
The code looks good otherwise. Perhaps two lines of documentation would help, something along the lines of

> To determine the actual position of the legend, check which corner (or center) of the legend is closest to the corresponding corner (or center) of the axes box.

**Comment by benoit9126** [comment]
I have corrected what you propose. Thanks.

What do you think of my remark on the anchors of the _center left_ and the _center right_ legend positions? I think when the legend is at the _center left_ position, the anchor should be _west_.

**Comment by nschloe** [comment]
Yes, I think you're right there. Would you fix that?

**Comment by nschloe** [comment]
Also, can you check what changed with the text_overlay test?

**Comment by benoit9126** [comment]
Yes, of course, I will. I do it tomorrow.

**Comment by benoit9126** [comment]
The difference in `text_overlay` is (obviously) the legend position. The selected legend position is `best` in the example so before my update, the legend was `upper right` by default and now, it is `upper left` as matplotlib does.
Here the old version of the `text_overlay` example ([acid_old_version.pdf](https://github.com/nschloe/matplotlib2tikz/files/912347/acid_old_version.pdf)) and here the new version 
([acid_new_version.pdf](https://github.com/nschloe/matplotlib2tikz/files/912348/acid_new_version.pdf)).

**Comment by benoit9126** [comment]
Two examples fail:
- the `best_legend_location`: here the [reference plot](http://chunk.io/f/c799255154344677bdefc19b1f03706a) and here the [new PDF plot](http://chunk.io/f/1deb85af647049dcb6b93afcd4a8bb1f).
- the `legends2`: here the [reference plot](http://chunk.io/f/5297aa66ba2e4676821ffb29eaa0e63b) and here the [new PDF plot](http://chunk.io/f/5412a350333f4743bee8faeea4db3411)

The first one looks good but the second one is ugly with the legend location 5 (right). I suggest to change the anchor 'west' to 'east' for this position too.

**Comment by nschloe** [comment]
>  I suggest to change the anchor 'west' to 'east' for this position too.

I agree.

(Also, I've just merged another PR that changed many phashes. Sorry for the conflict, but hopefully it's easily resolved.)

**Comment by benoit9126** [comment]
No problem for the other PR. :-)

[Here](http://chunk.io/f/cb6dc9fa40094c358e8c0cb90bc03b20) is the new `legend2` test with the new anchor position. It looks much better so I change the hashes.

**Comment by benoit9126** [comment]
Finally, it works! :+1: I have squashed my commits into three commits.

**Comment by nschloe** [comment]
Thanks for the nice contribution!

**Pull request merged**